src/flexmeasures_client/client.py

```diff
@@ -0,0 +1,287 @@
+from __future__ import annotations
+
+import asyncio
+import socket
+from dataclasses import dataclass
+from typing import Any, cast
+
+import async_timeout
+import pandas as pd
+from aiohttp.client import ClientError, ClientSession
+from yarl import URL
+
+from flexmeasures_client.response_handling import (
+    check_response,
+    check_content_type,
+    check_for_status,
+)
+CONTENT_TYPE_HEADERS = {
+    "Content-Type": "application/json",
+}
+API_VERSIOM = "v3_0"
+
+MAX_POLLING_STEPS: int = 10  # seconds
+POLLING_TIMEOUT = 200.0  # seconds
+REQUEST_TIMEOUT = 20.0  # seconds
+POLLING_INTERVAL = float(10)  # seconds
+
+
+
+@dataclass
+class FlexMeasuresClient:
+    """Main class for connecting to the FlexMeasures API"""
+
+    password: str
+    email: str
+    access_token: str = None
+    host: str = "localhost:5000"
+    scheme: str = "http" if "localhost" in host else "https"
+    ssl: bool = False if "localhost" in host else True
+    api_version: str = API_VERSIOM
+    path: str = f"/api/{api_version}/"
+    consumption_price_sensor: int = 3 #TODO find sensor and use sensor through API or set in config
+    reauth_once: bool = True
+
+    polling_step: int = 0
+    max_polling_steps: int = MAX_POLLING_STEPS  # seconds
+    polling_timeout: float = POLLING_TIMEOUT  # seconds
+    request_timeout: float = REQUEST_TIMEOUT  # seconds
+    polling_interval: float = POLLING_INTERVAL  # seconds
+    session: ClientSession | None = None
+
+    async def close(self):
+        await self.session.close()
+
+    async def request(
+        self,
+        uri: str,
+        *,
+        json: dict | None = None,
+        method: str = "POST",
+        path: str = path,
+        params: dict[str, Any] | None = None,
+        include_auth: bool = True,
+    ) -> tuple[dict, int]:
+        """Send a request to FlexMeasures.
+
+        Retries if:
+        - the client request timed out (as indicated by the client's self.request_timeout)
+        - the server response indicates a 408 (Request Timeout) status
+        - the server response indicates a 503 (Service Unavailable) status with a Retry-After response header
+
+        Fails if:
+        - the server response indicated a status code of 400 or higher
+        - the client polling timed out (as indicated by the client's self.polling_timeout)
+        """
+        url = self.build_url(uri, path=path)
+        print(url)
+
+        headers = await self.get_headers(include_auth=include_auth)
+        self.start_session()
+
+        self.polling_step = 0
```

Comment:
Great catch, this could be a nasty issue so need to think on how to do this. https://github.com/FlexMeasures/flexmeasures-client/issues/10

Feel free to correct the issue text and title if there are things that are not correct.